### ISSUE TITLE
What are lifecycle events for?

### ISSUE THREAD

**Issue by jorgebucaran** [hyperapp/hyperapp#717]
What are lifecycle events for and **can we do without them**? Below is each lifecycle event and its use cases. 

## `oncreate(element)`
1. Create fade/slide-in transitions when an element appears on the screen.
2. Add global event listeners, start timers, create websockets.
3. Fetch some data when the element appears on the screen.
4. Simulate local component state by storing data in the element.
5. Modify the element attributes, children or inner html (integration with third party libraries).
6. Set innerHTML.

### Alternatives
1. Store the initial style in the state, `style.opacity: 0` and use a tick subscription to create an interval. The interval will last **until** you go past/pass a threshold `style.opacity >= 1`. Update the state interactively at every interval `=> opacity+.2`. The same strategy can be used for fade/slide-out transitions.

2. Use a subscription for global event listeners, timers, websockets, etc.

3. Use the `init` function to set the initial state and immediately start an effect that will fetch your data and deliver it to another action. Inside that action you can set the next state. You could even set the next state and start another effect creating effectively a chain of effects.
    - Elements appearing or disappearing on the screen is a by-product of the state. The state changing is the result of user interacting with the DOM like clicks, etc., or external events, like global events, timers, url changes, geolocation changes, etc. Therefore, you can fetch data which will be "consumed" by some component via the same actions that update the state so that those components are displayed in the first place! 
    - Remember that actions can return not only the `=> nextState` but also start an effect `=> [nextState, someEffect]`.

4. [Custom Elements](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements)
      - Your custom elements can trigger [CustomEvents](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) and Hyperapp can respond to them via subscriptions (receiving). 
  
    - Conversely,  you can trigger effects that communicate with Custom Elements via Custom Events (sending/giving).

5. You can watch for changes being made to the DOM tree using [Mutation Observers](https://developers.google.com/web/updates/2012/02/Detect-DOM-changes-with-Mutation-Observers). It's not as subtle as using lifecycle events, but third party integrations are not subtle anyway. In fact, the result will be more maintainable code. Clean, declarative, pure Hyperapp code in one side, and imperative, hard-knocks code in the other side.

6. Provide better innerHTML support. Current support is questionable. For example: https://github.com/hyperapp/hyperapp/pull/688 /cc @frenzzy.

## `onupdate(element, oldProps)`
1. Modify the element attributes, children or inner html (integration with third party libraries), e.g., update a third party element.
2. Set innerHTML.

### Alternatives
1. Use mutation observers as described above.
2. Provide better innerHTML support. See (6) above.

## `onremove(element, done)`
1. Create fade/slide-out transitions by delaying the removal of the element at will.

### Alternatives
1. Use the state and subscriptions to update the state and animate elements and then remove the element instead of delaying the removal of the element. In other words, you have the state change happen _after_ the animation is complete, as opposed to take the matter in your own hands.

## `ondestroy`
1. Modify the element attributes, children or inner html (integration with third party libraries), e.g., destroy/remove a third party element.
2. Remove event listeners, clean timers/intervals, closing a websocket, etc.

### Alternatives
1. See using mutation observers or custom elements above.
2. See using subscriptions above.

---

Thoughts? 🤔

**Comment by whaaaley** [comment]
In HA1 I stopped using oncreate events for alternatives similar to what you described. I don’t like the view having control over listeners, animations, or fetching. In HA1 those things are best left to interop.

**Comment by SkaterDad** [comment]
My lifecycle uses are currently limited to embedding 3rd party widgets like Google Maps, and some very limited animations.

I would probably use the lifecycles for more animation stuff if more hooks were exposed, but I understand the desire to get rid of them to simplify the library code.

Using custom elements is something that Richard Feldman is always suggesting for people in elm, but I do have concerns about the polyfills due to low browser support.  Does anyone have experience with them, and know how many KB they add to the application bundle?  It seems intuitive to me that the devices that need the polyfill are the ones which also need the lightest/fastest javascript bundles.

**Comment by vladimirkosmala** [comment]
1. animations should be done with CSS
2. listeners and websockets should be done in model part
3. fetching: same as above, model knows when to fetch. Example: route change ? fetch things accordingly.
I understand that it can feel more confortable to have a `<list-users oncreate() {model.fetch()} .../>` though.

From personal experience, I use hooks to:
- listen scroll events for infinite scrolling
- bind with graphic libs
Custom Elements could replace them indeed.

> Elements appearing or disappearing on the screen is a by-product of the state. The state changing is the result of user interacting with the DOM like clicks, etc., or external events, like global events, timers, url changes, geolocation changes, etc.
Yes, definitely.
A _view_ should be drawn with a raw object and _model_ should be diagnosed with a Console alone as a raw object. It also helps for testing.

**Comment by jorgebucaran** [comment]
Thanks, @vladimirkosmala.

About:
>  ...listen scroll events for infinite scrolling

Listening to external/global events will work gracefully in V2 via subscriptions! 🎉  If you are familiar with Elm, you'll love it.

Essentially, there will be a Subscriptions API for implementers and your client code will look like something like this: (just an example)

```js
// With JSX
app({ 
  ...,
  subscribe: state => state.isTracking && <Geolocation.changes action={MyAction} />
})

// No JSX

app({ 
  ...,
  subscribe: state => state.isTracking && Geolocation.changes({ action: MyAction })
})
```

**Comment by jorgebucaran** [comment]
@SkaterDad Using custom elements is possibly the best solution to integrate a third party UI library, e.g., CodeMirror, etc. Another way is creating a vanilla wrapper communicates back and forth with Hyperapp using the Effects and Subscriptions API.

**Comment by zaceno** [comment]
@SkaterDad 

The v1 spec of custom elements (that is most similar to React components) currently needs polyfills on IE, Edge and FF. Mobile Safari & Chrome for Android work. So I think you’re good. https://caniuse.com/#search=Custom

**Comment by whaaaley** [comment]
@vladimirkosmala Not all animations can be accomplished with _just_ CSS. Sometimes you need dynamic values, like an element location or size, then apply those values within CSS inline.

**Comment by selfup** [comment]
I like the idea of subscriptions in 2.0 for frontloading listeners, or just listeners in general.

Excited to see where this goes 🙏

**Comment by mrozbarry** [comment]
I'm currently using `oncreate` a lot with `@hyperapp/router`, since it feels more natural to put fetching in an oncreate action, rather than adding an external `location.subscribe` method with a large switch statement.

Especially in terms of SPA, there can be multiple links to navigate to a page, so adding onclick handlers to all of those links wouldn't be very efficient and would be hard to manage.

**Comment by atomiks** [comment]
> somehow react to the element being removed and inject your own logic for the removal. Prevent it from being removed directly, apply the fade-out class, and after 300ms remove it properly. This is my preferred strategy

I'm having trouble how this would be possible without `onremove` and the `done` callback. With MO, it detects when it has been removed but by then it's already too late?

**Comment by zaceno** [comment]
@atomiks You're right. It's not entirely straightforward. But because I have the parent (which the MO is observing), and the child (which I get in the mutation callback), I can basically just put the removed element back into the parent so I can remove it myself later. 😜 

But that means I need to make sure I don't react to the mutation of putting it back in... so there are some contortions to go through... I'm not sure it's 100% safe the way I'm doing it but it seems to be working alright.

**Comment by jorgebucaran** [comment]
> @mrozbarry since it feels more natural to put fetching in an oncreate action, rather than adding an external location.subscribe method with a large switch statement.

V2 will solve this nicely with subscriptions, this is an easy one.

---

> @whaaaley Not all animations can be accomplished with just CSS. Sometimes you need dynamic values, like an element location or size, then apply those values within CSS inline.

Perhaps a `@hyperapp/transitions`-like library built on Mutation Observers could exist to handle more complicated cases and simplify keeping the animation state in the `state`.

---

> @zaceno One problem though, is how to apply the exit-styles for an element that should fade out?

Everything is done with the state. See: https://github.com/mdgriffith/elm-style-animation/tree/3.5.5

I tried my best to explain it already so I won't repeat myself. See (3) in Alternatives to `oncreate` [above](https://github.com/hyperapp/hyperapp/issues/717#issue-337438224).

**Comment by vladimirkosmala** [comment]
Note on Custom Elements, to communicate between them and hyperapp, we would need to pass data though attributes.
Example: `h('UserElement', {'data-name': 'Brian'})`

The problem is on structures, sending JSON is heavy.
Example: `h('UsersElement', {'data-users': '[{"name": "Brian", }, ...]'})`

Current hooks are efficient here:
`h('div', {oncreate: (element) => bootExternalLib(element, model.state.value)})`

What do you think?

**Comment by jorgebucaran** [comment]
@vladimirkosmala For communicating between custom elements and Hyperapp and vice-versa I am only considering using the new Subscriptions and Effects API, which is similar to how ports work in Elm.

**Comment by SkaterDad** [comment]
@vladimirkosmala It may be worth looking at what Preact does.   https://custom-elements-everywhere.com/ has descriptions of how the various frameworks handle these problems, and test cases for library authors to check compliance.  Many of the libraries set properties when possible, and string attributes as a fallback.

@jorgebucaran Would it not be possible to interact with custom elements just like regular ones?  I thought that was the whole point of them.  I'd much prefer something like below, to avoid having to wire up subscriptions.

```jsx
<map-component lat={90.1} lon={-40} onchange={(event)=> {}} />
```

**_Site note:_** I did a little research, and the custom elements polyfill appears to be ~3 KB min+gzip (according to X-Tag's website).  Pretty small, but not needed if you keep the lifecycles.

**Comment by jorgebucaran** [comment]
> @SkaterDad Would it not be possible to interact with custom elements just like regular ones?  

Yes, you are right. You don't need to use effects or subscriptions with custom elements, but you could.

**Comment by zaceno** [comment]
I posted this over in #720 but realized it makes even more sense in this thread:

--

I tossed @hyperapp/transitions into a pot with MutationObserver and customElements, and let it all boil together into a framework agnostic custom element: <transition-group> with a similar behavior as the <transition> tag of Vue.js.

The implementation is definitely not elegant by any means (marrying the contorsions of @hyperapp/transitions with the cludgy lifecycle-hook-emulation of mutation observers). It's probably the least fun programming I've had in a long time. But it works.

Here are the examples from @hyperapp/transitions ported to use my <transition-group> instead:

Carousel
15-Puzzle
Toasts
And here's the codepen with the "implementation" 🤢 : https://codepen.io/zaceno/pen/qKwXXK?editors=0010

**Comment by dmitrykurmanov** [comment]
_moved from the duplicating #736_

> What about using popular libraries? I think that it is a real use case. For example standard jQuery datepicker.
> I've created an example with the datepicker and hyperapp1 with oncreate event
> https://jsfiddle.net/badgoto/tfoxqd6h
> How can I achive this in hyperapp2 ?

@jorgebucaran:

> I eventually want to remove lifecycle events from Hyperapp, but I'm most likely going to keep them for HAV2.

Thanks for the answer. So what about support different third party widgets like the datepicker above? How will they work without events? In fact for that we need to know when a node appears in the dom. May be I've missed something.

**Comment by jorgebucaran** [comment]
I gave it a LOT of thought and I am _not_ going to remove lifecycle events from Hyperapp 2, 3 or 4. Not unless a clear, impeccable alternative is discovered or proposed. Until then, lifecycle events are going to stay.

We can still create _💯_ percent functional apps by avoiding them and I encourage everyone to find alternatives to the problems we currently solve with them e.g., always using a state-first approach, [using custom elements](https://github.com/jorgebucaran/superfine/issues/135), mutation observers, etc. (https://github.com/hyperapp/hyperapp/issues/717#issue-337438224)

HAV2 aims to reduce the amount of imperative code we write to create apps. Embrace this vision and build purely-functional, state-first, declarative apps, but rest assured  that lifecycle events will be available when you need a way to escape from wonderland.

---

To answer @dmitrykurmanov's question: 
> So what about support different third party widgets like the datepicker above? 
> How will they work without events?

The answer is you can't **without** lifecycle events. If you want a purely functional datepicker that is consistent with the rest of your Hyperapp application you need to build it from scratch.

**Comment by sergey-shpak** [comment]
Hi guys,
I've reviewed and grouped most common lifecycle events usages:
1. Start Listeners, Timers, Sockets / Fetch some data
2. Element Animations / Element Modifications

#### Starting Listeners / timers / sockets or fetching some data 
HyperappV1 actions are wrapped with `dispatch` function, so calling such actions from anywhere will trigger state update (and view redraw as result). That's why it's so easy to update state from lifecycle events listeners (just to call action).
HyperappV2 actions are pure and simple, to update state you should manually pass action to `dispatch` function. Lifecycle event listeners are not passed to `dispatch` function **as result, lifecycle events [listeners cannot update state](https://github.com/jorgebucaran/hyperapp/blob/V2/src/index.js#L448)**. (they can call action, but the state won't be updated), so **there is no practical sense to start timers or fetch some data with lifecycle events**. (Subscriptions/effects were specially designed for such purposes)

#### Element Animations / Element Modifications
Most animations should be done with css. 
Each element modification should be reflected as vNode modification. 
I agree on some small exceptions, like: fadein/fadeout animation, addition html markup, etc
______

Lifecycle events are handy and simple tools to achieve immediate result, but the scope of their usage is too limited. If we are taking into account interoperability / integrations, lifecycle events can be replaced with more powerful patterns like subscriptions/effects.
______

#### The way to handle onCreate/onUpdate/onRemove without lifecycles
[Gist](https://gist.github.com/sergey-shpak/f16cca595c05387928fb893560bd28ec) and [Real working example](https://codepen.io/sergey-shpak/pen/vVgEbG)

```javascript
import DOM from 'subcriptions/dom'

const lifecycle = (state, element, params) => {
  // handle dom element changes
}

const conainer = document.body
app({
  init: {},
  view: state => <div>
    <div id="someid" >div</div>
  </div>,
  subscriptions: state => [
    DOM.observe({
      select: { id: 'someid' },
      target: container,
      action: lifecycle
    })
  ],
  container
})

```

**Comment by zaceno** [comment]
@sergey-shpak That is one very clever approach! 👍 

...But I don't consider it a complete answer/replacement to lifecycle events.

- For one thing, if you have many components using lifecycle events today, each one will need it's own  custom row in the subscriptions. Wiring everything together could become quite messy and "boilerplaty". 

- One thing that is often forgotten about when talking about using mutationobserver instead of lifecycle events for transition-animations: Exit-animations.  It's not possible to apply a css transition to an element that has already been removed. This is one (perhaps the only) reason why having `onRemove` is so great. 

- Also, this approach doesn't work (I don't think... 🤔 ) for another of my favorite lifecycle use-cases: Rendering in canvas-elements using `onUpdate`

I'm sure there are solutions to all the various uses of lifecycle-events out there in the wild today. But those solutions are sure to be a lot more complicated, verbose and messy, to wire things up (or if the complexity is encapsulated in a library,  it will necessarily be quite specific). In my opinion lifecycle-events solve a general problem of giving access to manipulating the raw DOM, in a very clean, compact and developer-friendly way.

On the other hand, perhaps that point is less relevant to v2, since v2 isn't trying to avoid DIY-wiring / boilerplate anyway.

**Comment by ivan-kleshnin** [comment]
```
DOM.observe({
  select: { id: 'someid' },
  target: container,
  action: lifecycle
})
```

@sergey-shpak Can you target repeating elements of the same type (where no ids are possible) 
with this approach?

**Comment by sergey-shpak** [comment]
@ivan-kleshnin since we are only discussing lifecycles, the example provided as proof of concept, final Mutation Sub should have more generic querying approach

**Comment by jorgebucaran** [comment]
@sergey-shpak 👍

**Comment by jklarenbeek** [comment]
Just on a little side note, CSS3 transitions are very primitive and clumbsy. I'd rather use libraries like velocityjs to do my animations, eventhough I use animate.css once in a while too.
Another thing I like to say is that since hyperapp2 is so totally different from v1, wouldn't it be better to rename hyperapp to hyperapp2 so there is no confusion of what hyperapp paradigm one is using...?

**Comment by jorgebucaran** [comment]
> @jklarenbeek ...wouldn't it be better to rename Hyperapp to Hyperapp2 so there is no confusion of what Hyperapp paradigm one is using...?

I couldn't afford that, but really, I don't think it's necessary as I will explain below.

> ...since V2 is so totally different from V1

This is a misconception. V1 and V2 are more alike than different. V1 can be used in a way which is  similar to how V2 will be used through [hyperapp-fx](https://github.com/okwolf/hyperapp-fx). The are differences, but there are not enough differences to warrant a product rename.

**Comment by mindplay-dk** [comment]
The main thing that makes the create/remove events interesting to me, is the ability to animate adding/removing list items. (To me, this is one of the only valid, useful cases for animation.)

It seems to be the one case for which you haven't presented an alternative?

Having to put animation state in the actual model would seem problematic to me - for example, if you put animation state in the items of a to-do list, this would affect logic like counting and filtering, which would be less than elegant.

What's nice about life cycle events, is you can deal with something purely UI related in the domain of the UI (DOM) rather than having to model it in your application domain.

If there's a practical alternative to that in V2, I'm not seeing it yet.

**Comment by jorgebucaran** [comment]
@mindplay-dk I recently updated this issue, see that I mentioned three distinct ways for creating animations without lifecycle events: [using CSS](#animations-using-css) ([CodePen](https://codepen.io/anon/pen/moEpJQ)), [using effects](#animations-using-fx) ([CodePen](https://codepen.io/anon/pen/oVxGvv)), and [using subscriptions](#animations-using-subs) ([CodePen](https://codepen.io/anon/pen/moeZXg)). 

Effects can be a practical way to bridge the gap between V2's purely declarative style and V1's mostly declarative, sometimes imperative mix. Of course, effects are not limited to wrapping imperative DOM operations; you can use them to wrap almost everything. [Read more](https://github.com/jorgebucaran/hyperapp/issues/750).

I don't have any practical experience using subscriptions to model animations, and I gather neither do you, so we should get some real experience first and then draw our conclusions. Many Elmers do it that way (see this talk by Mathew Griffith on [Elm Style Animation](https://www.youtube.com/watch?v=DsDwYqsLU3E)). And I think it's logical in a system where the state is the only source of truth.

**Comment by jorgebucaran** [comment]
@mindplay-dk I recently updated this issue, see that I mentioned three distinct ways for creating animations without lifecycle events: [using CSS](#animations-using-css) ([CodePen](https://codepen.io/anon/pen/moEpJQ)), [using effects](#animations-using-fx) ([CodePen](https://codepen.io/anon/pen/oVxGvv)), and [using subscriptions](#animations-using-subs) ([CodePen](https://codepen.io/anon/pen/moeZXg)). 

## Using Effects

Effects can be a practical way to bridge the gap between V2's purely declarative style and V1's mostly declarative, sometimes imperative mix. Of course, effects are not limited to wrapping imperative DOM operations; you can use them to wrap almost everything. [Read more](https://github.com/jorgebucaran/hyperapp/issues/750).

First, we describe our view and tell Hyperapp to dispatch some action when the user clicks a button.

```jsx
<button onClick={Shuffle}>Shuffle</button>
```

`Shuffle` returns a `[state, Effect]` tuple—that's how we tell Hyperapp to run side effects. `Flip` is the function that returns the effect. Nothing prevents you from writing: `[myFlipEffect, props]`, but I prefer `Flip(props)`.

```jsx
const Shuffle = state => [
  { ...state, items: shuffle(state.items) },
  Flip(document.querySelectorAll(".cell"))
]
```

## Using Subscriptions

I don't have any practical experience using subscriptions to model animations, and I gather neither do you, so we should get some real experience first and then draw our conclusions. Many Elmers do it that way (see this talk by Mathew Griffith on [Elm Style Animation](https://www.youtube.com/watch?v=DsDwYqsLU3E)). And I think it's logical in a system where the state is the only source of truth.

**Comment by jorgebucaran** [comment]
@mindplay-dk I recently updated this issue, see that I mentioned three distinct ways for creating animations without lifecycle events: [using CSS](#animations-using-css) ([CodePen](https://codepen.io/anon/pen/moEpJQ)), [using effects](#animations-using-fx) ([CodePen](https://codepen.io/anon/pen/oVxGvv)), and [using subscriptions](#animations-using-subs) ([CodePen](https://codepen.io/anon/pen/moeZXg)). 

## Using Effects

Effects can be a practical way to bridge the gap between V2's purely declarative style and V1's mostly declarative, sometimes imperative mixed style. Of course, effects are not limited to wrapping imperative DOM operations; you can use them to wrap almost everything. [Read more](https://github.com/jorgebucaran/hyperapp/issues/750).

First, we describe our view and tell Hyperapp to dispatch some action when the user clicks a button.

```jsx
<button onClick={Shuffle}>Shuffle</button>
```

`Shuffle` returns a `[state, Effect]` tuple—that's how we tell Hyperapp to run side effects. `Flip` is the function that returns the effect. Nothing prevents you from writing: `[myFlipEffect, props]`, but I prefer `Flip(props)`.

```jsx
const Shuffle = state => [
  { ...state, items: shuffle(state.items) },
  Flip(document.querySelectorAll(".cell"))
]
```

## Using Subscriptions

I don't have any practical experience using subscriptions to model animations, and I gather neither do you, so we should get some real experience first and then draw our conclusions. Many Elmers do it that way ([watch this talk](https://www.youtube.com/watch?v=DsDwYqsLU3E) by Mathew Griffith on [Elm Style Animation](https://github.com/mdgriffith/elm-style-animation)). And I think it's logical in a system where the state is the only source of truth.

**Comment by zaceno** [comment]
@jorgebucaran I don't think you answered @mindplay-dk 's concern. What I took from his comment is this: "using the *state* to manage fade out animations from a *list*, will be very messy. The logic seems impossible to encapsulate in a user friendly way"

My response to that is: yes, you are correct. *If* `onRemove` could be used reliably across browsers,  without the need for doing tricks to the layout to removed elements, it would be a very elegant solution to the "fade-out elements from a list" problem.

I've tried and tried and finally given up on `onRemove`. I don't think there's any way it can be made to work reliably across browsers (and across Hyperapp updates). Also, there's the wierdness of needing to compensate for elements changing position when removed from the DOM.

So to summarize: yes it's a problem, no it hasn't been solved yet. `onRemove` looks like an answer, but is not. (As far as I'm able to tell)

**Comment by mindplay-dk** [comment]
> I mentioned three distinct ways for creating animations without lifecycle events

@jorgebucaran I reviewed those before posting my comment - they do not address the problem of adding/removing items; other than by complicating the app model with UI state - by which I mean, state that is exclusively relevant in terms of UI update and/or has not relevance to the application domain.

> So to summarize: yes it's a problem, no it hasn't been solved yet. onRemove looks like an answer, but is not. (As far as I'm able to tell)

@zaceno I don't fully understand the problem (or why it's browser-dependent) but if `onremove` is not the answer and conceptually doesn't "work", then I'm not against removing it.

But element-level life-cycle events were the original thing that attracted me to Superfine - I haven't done much with Hyperapp yet, but can't see any of the demonstrated alternatives as a substitute. (I'm not interested in the kinds of animation demonstrated in the examples - I find most animation in UI to be a bit gimmicky; for the most part, the only commonly useful/meaningful use-case for animation in most UI is additions/removals, because they actually help the user understand what's happening.)

So I hope to see something else introduced at some point to address these requirements.

On that note, UI state pertaining to animation during additions/removals is sort of a corner case that happens to correlate with the element life-cycle - in my optic, UI state is a broader category of problems that could include control state (for example, the state of a custom drop-down control, current month in a date-picker, etc.) something that still isn't supported by this model. Maybe something to think about, if perhaps there's a single concept that could cover all of these use-cases? No idea.

(and I know, "you can do that with custom elements", but many of us still have to ship for IE, so...)

**Comment by SkaterDad** [comment]
@mindplay-dk I've adapted the "Effects" sudoku animation example to animate list inserts/removals.   Might not be 100% what you're after, but it works pretty well.
https://codepen.io/SkaterDad/pen/RdRXXX

@jorgebucaran Thanks for posting the new examples.  I particularly appreciated the Effects example, and it makes a lot of sense in hindsight.

**Comment by mindplay-dk** [comment]
@SkaterDad nice! though removals are somewhat abrupt - presumably because the element is actually removed immediately during the update? I don't know how you'd get around that without deferred removals - you don't have the item in the model anymore, so an update will render without the resulting VNode from that item, and without deferred removals, well.... (?)

**Comment by jorgebucaran** [comment]
@SkaterDad Instead of animating the element leaving, animate the list. It's so zen. 
 
@mindplay-dk Animate hiding element and remove it when the transition is done?

https://codepen.io/anon/pen/RdGzpp
```jsx
const ItemRemoved = (state, itemId) => ({
  ...state,
  items: state.items.filter(item => item.id !== itemId)
})

const RemoveItem = (state, itemId) => [
  {
    ...state,
    items: state.items.map(item =>
      item.id === itemId ? { ...item, isRemoving: true } : item
    )
  },
  Time.delay({ duration: 500, action: [ItemRemoved, itemId] })
]

// ...
<ul>
  {state.items.map(item => (
    <li
      style={{
        transition: "all .5s",
        transform: item.isRemoving && "scale(0)"
      }}
      onClick={[RemoveItem, item.id]}
      key={item.id}
    >
      {item.text}
    </li>
  ))}
</ul>
```

**Comment by jorgebucaran** [comment]
> @SkaterDad I've adapted the "Effects" sudoku animation example to animate list inserts/removals. 

Instead of animating the element leaving, animate the list. It's so zen. 
 
> @mindplay-dk I don't know how you'd get around that without deferred removals - you don't have the item in the model anymore...

Animate hiding element and remove it when the transition is done ([CodePen](https://codepen.io/anon/pen/RdGzpp))? 🤔 

```jsx
const ItemRemoved = (state, itemId) => ({
  ...state,
  items: state.items.filter(item => item.id !== itemId)
})

const RemoveItem = (state, itemId) => [
  {
    ...state,
    items: state.items.map(item =>
      item.id === itemId ? { ...item, isRemoving: true } : item
    )
  },
  Time.delay({ duration: 500, action: [ItemRemoved, itemId] })
]

// ...
<ul>
  {state.items.map(item => (
    <li
      style={{
        transition: "all .5s",
        transform: item.isRemoving && "scale(0)"
      }}
      onClick={[RemoveItem, item.id]}
      key={item.id}
    >
      {item.text}
    </li>
  ))}
</ul>
```

**Comment by mindplay-dk** [comment]
@SkaterDad yeah, but... you're polluting the model with `isRemoving` - and the item isn't actually removed until after the UI is done updating. If for example you had to count the items in the list, you'd have to filter that ones with `isRemoving`. It complicates the model with UI details. This is what I didn't want, so... not really my idea of Zen 😐

**Comment by jorgebucaran** [comment]
@mindplay-dk I was referring to Mike's pen. I never suggested the example I created was zen. I still don't know if there's a simpler way to describe a delayed removal with what we have at the moment. 

My new guideline going forward is to remove all the escape hatches until good solutions are revealed to me, that's one of the reasons lifecycle events are going away. In the future, we may provide an entirely different solution to this problem. In a way, lifecycle events may return, but not exposed directly to you.

If you were to design this purely declarative API, how would it look like?

**Comment by SkaterDad** [comment]
@mindplay-dk One approach to animate the "removed" element is to clone it before the next animation frame, and insert it back where it was.

Effects are called during that time, conveniently.

Here is an updated Codepen using this technique.  The way I've handle it is a bit clunky, but I'm sure you can find ways to clean it up that work for your projects.
https://codepen.io/SkaterDad/pen/YgpNrp

**Comment by mindplay-dk** [comment]
> If you were to design this purely declarative API, how would it look?

I have no idea. V1 was easy to understand, but I still have trouble reading the V2 examples - hopefully understanding will come with better documentation and examples eventually.

Looks like you had some ideas kicking around in #766 though?

**Comment by apla** [comment]
I have two examples of lifecycle events which is not possible to implement using effects. They closely related to the DOM state/events/methods.

1. Video player. We have digital signage solution build on hyper app v1 and there is no way to start video conditionally without using lifecycle events. Our code code looks like this:

```javascript
function onVideoTagUpdate (element, oldAttributes) {
  if (element.parentNode.classList.contains ('active')) {
    element.play()
  }
}

function onVideoEnded ({target}) {
  dispatch ('widget.finished');
}

export default function VideoComponent ({widget, currentWidgetId}, children) {
  return <div class={widget.id === currentWidgetId ? 'active' : ''}>
    <video
      onended={onVideoEnded}                    
      onupdate={onVideoTagUpdate}
      muted="true"
      playsinline="true"
      preload="metadata"
      src={widget.el.currentSrc}
    />
  </div>;
}
```

Changing autoplay attribute just not works for some browser combinations after video already preloaded. We need to start only one video at precise time and know when it ends.

2. For the same solution, we have marquee, which moves strings one after another. Strings loaded from server and we can render it offscreen, then move using css animation. Using lifecycle events we measuring string length to calculate time delay between current and next string, then setting class to activate css animation. We cannot render all strings into one marquee because they changing at runtime and  for our screen 5000 pixels wide all strings rendered will fit in between 100 000 and 400 000 pixels, which normal browser cannot animate nor survive. 

As a solution, we could recreate string rendering and measure at controller side, but, unfortunately, animation which should run for about 5 seconds, can end in much later and we need to adjust it accordingly. Also, we can transmit element state to controller, but for V2 there is no way to get the rendered element, so we need to implement id generator and all this looks like pile of hacks.

**Comment by sergey-shpak** [comment]
@apla, I've implemented hyperapp#2 lifecycle events helper: 
https://gist.github.com/sergey-shpak/c1e0db3d52019eecb0b5717e8cbf00ad
it looks pretty straightforward and clear, but please use it carefully

**Comment by zaceno** [comment]
@apla You could also consider implementing the `VideoComponent` as a [custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements).
